File: frontend/src/api/actividades.ts
import { get, patch, post } from './client';

export type ActivityStatus = 'PENDIENTE' | 'EN_PROCESO' | 'COMPLETADA';
export type ActivityDateType = 'PROGRAMADA' | 'LIMITE';
export interface Actividad { id_actividad: string; nombre: string; descripcion: string; responsable_id: string; responsable: string | null; tipo_fecha: ActivityDateType; fecha_objetivo: string; estado: ActivityStatus; persona_id: string | null; persona: string | null; cedula: string | null; area: string | null; registrado_por: string | null; fecha_creacion: string | null; fecha_finalizacion: string | null; vencida: boolean; version: number; }
export interface ActivityOptions { usuarios: { id: string; nombre: string }[]; personas: { id: string; nombre: string; cedula: string | null; area: string | null }[]; }
export interface ActivityPayload { nombre: string; descripcion: string; responsable_id: string; tipo_fecha: ActivityDateType; fecha_objetivo: string; estado: ActivityStatus; persona_id: string | null; }
export function listActivities(params: URLSearchParams): Promise<{ items: Actividad[]; total: number; limite: number; offset: number }> { return get(`/actividades?${params}`); }
export const activityOptions = (): Promise<ActivityOptions> => get('/actividades/opciones');
export const createActivity = (data: ActivityPayload): Promise<Actividad> => post('/actividades', data);
export const updateActivity = (id: string, data: ActivityPayload & { expected_version: number }): Promise<Actividad> => patch(`/actividades/${id}`, data);
export const archiveActivity = (id: string, expected_version: number, motivo: string): Promise<Actividad> => post(`/actividades/${id}/archivar`, { expected_version, motivo });

File: backend/app/models/actividades.py
from sqlalchemy import CheckConstraint, ForeignKey, Index, String
from sqlalchemy.orm import Mapped, mapped_column

from app.models.security import Base, MetadatosComunes


class Actividad(MetadatosComunes, Base):
    __tablename__ = "actividades"
    __table_args__ = (
        CheckConstraint("version >= 1", name="version_positive"),
        Index("ix_actividades_estado_fecha_objetivo", "estado", "fecha_objetivo"),
    )

    id_actividad: Mapped[str] = mapped_column(String, primary_key=True)
    nombre: Mapped[str] = mapped_column(String, nullable=False)
    descripcion: Mapped[str] = mapped_column(String, nullable=False)
    responsable_id: Mapped[str] = mapped_column(ForeignKey("usuarios.id_usuario"), index=True, nullable=False)
    tipo_fecha: Mapped[str] = mapped_column(String, nullable=False)
    fecha_objetivo: Mapped[str] = mapped_column(String, nullable=False)
    estado: Mapped[str] = mapped_column(String, nullable=False, default="PENDIENTE")
    persona_id: Mapped[str | None] = mapped_column(ForeignKey("personas.id_persona"), index=True)
    creado_por_id: Mapped[str] = mapped_column(ForeignKey("usuarios.id_usuario"), index=True, nullable=False)
    fecha_finalizacion: Mapped[str | None] = mapped_column(String)
